bson-bench/src/task.ts

```diff
@@ -0,0 +1,179 @@
+import { once } from "events";
+import { writeFile } from "fs/promises";
+import * as path from "path";
+
+import {
+  type BenchmarkResult,
+  type BenchmarkSpecification,
+  type ErrorMessage,
+  Package,
+  type PerfSendResult,
+  type ResultMessage,
+} from "./common";
+
+export class Task {
+  result: BenchmarkResult | undefined;
+  benchmark: BenchmarkSpecification;
+  taskName: string;
+  /** @internal */
+  children: ChildProcess[];
+  /** @internal */
+  hasRun: boolean;
+
+  constructor(benchmarkSpec: BenchmarkSpecification) {
+    this.result = undefined;
+    this.benchmark = benchmarkSpec;
+    this.children = [];
+    this.hasRun = false;
+
+    this.taskName = `${path.basename(this.benchmark.documentPath, "json")}_${
+      this.benchmark.operation
+    }_${this.benchmark.library}`;
+  }
+
+  getResults(): PerfSendResult {
+    if (!this.hasRun)
+      throw new Error("cannot write results if benchmark has not been run");
+    if (!this.result) throw new Error("benchmark failed; no results to write");
+
+    const { durationMillis, documentSizeBytes } = this.result;
+    const calculateThroughputMBps = (durationMillis: number, bytes: number) =>
+      bytes / durationMillis / 1000;
+    const throughputMBps = durationMillis.map((d) =>
+      calculateThroughputMBps(d, documentSizeBytes),
+    );
+    throughputMBps.sort();
+
+    // Calculate summary statistics
+    const meanThroughputMBps =
+      throughputMBps.reduce((prev, thrpt) => thrpt + prev, 0) /
+      throughputMBps.length;
+    const medianThroughputMBps = throughputMBps[throughputMBps.length / 2];
+    const maxThroughputMBps = throughputMBps[throughputMBps.length - 1];
+    const minThroughputMBps = throughputMBps[0];
+
+    const convertOptions = (o: any): any => {
+      const output = Object.create(null);
+      for (const key in o) {
+        // boolean options
+        switch (key) {
+          case "promoteValues":
+          case "promoteBuffers":
+          case "promoteLongs":
+          case "bsonRegExp":
+          case "allowObjectSmallerThanBufferSize":
+          case "useBigInt64":
+          case "evalFunctions":
+          case "cacheFunctions":
+          case "checkKeys":
+          case "ignoreUndefined":
+          case "serializeFunctions":
+            output[key] = Number(o[key]);
+            break;
+          // numeric options
+          case "index":
+            output[key] = o[key];
+            break;
+          // special cases
+          case "validation":
+            output["utf8Validation"] = /^bson-ext/.test(this.benchmark.library)
+              ? 1
+              : typeof o[key] === "boolean"
+              ? Number(o[key])
+              : 1;
+            break;
+          default:
+            output[key] =
+              typeof o[key] === "boolean" || typeof o[key] === "number"
+                ? Number(o[key])
+                : -1;
+        }
+      }
+
+      return output;
+    };
```

Comment:
This is here to ensure we can collect our agrs since perf.send only supports options with numeric values in the `info.args` field